### ISSUE TITLE
NoSuchBucket error during Toil context manager exit

### ISSUE THREAD

**Issue by hannes-ucsc** [BD2KGenomics/toil#921]
I have the feeling this happens while `__exit__` is invoked with a primary exception. The primary exception is shadowed by the secondary exception that results from not being able to delete the job store. The robust deletion PR might improve this but we should also log the primary exception eagerly in `__exit__`.

**Comment by JakeNarkizian** [comment]
@hannes-ucsc Yes, robust clean will ignore any exceptions of this nature completely. How about logging any exceptions caught in ```AbstractJobStore._deleteJobStore```(the main body of each ```cleanJobStore``` and ```deleteJobStore``` methods) at the debug level?

**Comment by hannes-ucsc** [comment]
Yes, secondary exceptions that you swallow should be logged at debug level.

But I am concerned about the primary exception. Maybe there should be a general `except:` clause around `Toil.__exit__` that catches any exception, and depending on whether `__exit__` was called with a primary exception, logs it or reraises it. It should log the secondary exception if a primary exception was passed in. It should reraise the secondary exception if no primary exception was passed in.

**Comment by hannes-ucsc** [comment]
If the robust clean PR addresses this, this issue should be referenced in that PR's title and in the body. Then Waffle will stack the two.

**Comment by hannes-ucsc** [comment]
@JakeNarkizian, did you see [my previous comment](https://github.com/BD2KGenomics/toil/issues/921#issuecomment-226684843)?

**Comment by JakeNarkizian** [comment]
@hannes-ucsc Yes, thanks for the reminder. I haven't addressed this in #900, yet. 

Robust clean ignores exceptions raised due to corrupted job stores components, and it is reasonably implied what components were present at the time of cleaning in the log, but specific exceptions like ```NoSuchBucket``` aren't logged explicitly. The Toil context manager also doesn't log anything differently. 

I definitely can resolve this issue as well if you think it is better since I am touching code around there anyway, but I'm slightly hesitant to take on anymore because of how big this PR is already. 

It currently addresses 7 tracked issues, and a few others that don't have tickets (like using the general retry logic you've implemented for google, refactoring job store string to job store locator, I'm sure I'm blanking on more than one at the moment...). It also refactors quite a bit of each of the four currently supported job stores because of the need for class methods for each job store component. I've been delayed a few times by merge conflicts due to the sheer amount of code I've had to change in the job stores. 

That said, it might be worth addressing this issue as well since--like i mentioned above--the code's already touched. What do you think?

**Comment by JakeNarkizian** [comment]
**tl;dr** 

I haven't addressed it yet.
It might be wise to since I'm touching the code in common, but the PR is already really big.
what do you think?

**Comment by hannes-ucsc** [comment]
You probably mean #916, not #900. Anyways, I thought that #916 would solve this automatically. Maybe I'm not understanding this correctly.

So #916 is not going to make the cut for 3.2.0. Can you break #916 up into smaller PRs?

**Comment by JakeNarkizian** [comment]
Yes, sorry #916. 
I can break #916 into separate PRs but each branch is going to have to build off the last in a somewhat messy way since its changes are very closely related, I'd like to run it by you before I start the process.

Did you had a chance to glance at #916's loadBucket and createBucket methods in AWS? Perhaps we can discuss this when we talk about those methods.

**Comment by hannes-ucsc** [comment]
Punting for next sprint.
